docs/design_proposals/multi_environment_proposal.md

```diff
@@ -0,0 +1,334 @@
+- GitHub project: https://github.com/orgs/prefix-dev/projects/10
+
+## Real world example use cases
+??? tip "Polarify test setup"
+    In `polarify` they want to test multiple versions combined with multiple versions of polars.
+    This is currently done by using a matrix in GitHub actions.
+    This can be replaced by using multiple environments.
+
+    ```toml title="pixi.toml"
+    [project]
+    name = "polarify"
+    # ...
+    channels = ["conda-forge"]
+    platforms = ["linux-64", "osx-arm64", "osx-64", "win-64"]
+
+    [tasks]
+    postinstall = "pip install --no-build-isolation --no-deps --disable-pip-version-check -e ."
+
+    [dependencies]
+    python = ">=3.9"
+    pip = "*"
+    polars = ">=0.14.24,<0.21"
+
+    [feature.py39.dependencies]
+    python = "3.9.*"
+    [feature.py310.dependencies]
+    python = "3.10.*"
+    [feature.py311.dependencies]
+    python = "3.11.*"
+    [feature.py312.dependencies]
+    python = "3.12.*"
+    [feature.pl017.dependencies]
+    polars = "0.17.*"
+    [feature.pl018.dependencies]
+    polars = "0.18.*"
+    [feature.pl019.dependencies]
+    polars = "0.19.*"
+    [feature.pl020.dependencies]
+    polars = "0.20.*"
+
+    [feature.test.dependencies]
+    pytest = "*"
+    pytest-md = "*"
+    pytest-emoji = "*"
+    hypothesis = "*"
+    [feature.test.tasks]
+    test = "pytest"
+
+    [feature.lint.dependencies]
+    pre-commit = "*"
+    [feature.lint.tasks]
+    lint = "pre-commit run --all"
+
+    [environments]
+    pl017 = ["pl017", "py39", "test"]
+    pl018 = ["pl018", "py39", "test"]
+    pl019 = ["pl019", "py39", "test"]
+    pl020 = ["pl020", "py39", "test"]
+    py39 = ["py39", "test"]
+    py310 = ["py310", "test"]
+    py311 = ["py311", "test"]
+    py312 = ["py312", "test"]
+    ```
+
+    ```yaml title=".github/workflows/test.yml"
+    jobs:
+      tests:
+      name: Test ${{ matrix.environment }}
+      runs-on: ubuntu-latest
+      strategy:
+        matrix:
+          environment:
+            - pl017
+            - pl018
+            - pl019
+            - pl020
+            - py39
+            - py310
+            - py311
+            - py312
+      steps:
+        - uses: actions/checkout@v4
+        - uses: prefix-dev/setup-pixi@v0.5.0
+          with:
+            # already installs the corresponding environment and caches it
+            environments: ${{ matrix.environment }}
+        - name: Install dependencies
+          run: |
+            pixi run --env ${{ matrix.environment }} postinstall
+            pixi run --env ${{ matrix.environment }} test
+    ```
+??? tip "Test vs Production example"
+    This is an example of a project that has a `test` feature and `prod` environment.
+    The `prod` environment is a production environment that contains the run dependencies.
+    The `test` feature is a set of dependencies and tasks that we want to put on top of the previously solved `prod` environment.
```

Comment:
> on top

Either this wording is confusing/imprecise, or we are thinking about a different use case.

Generally speaking the technical process to come up with the `test` environment must be to solve the `test` environment first, and then _remove_ stuff from it to come up with the `prod` environment. You cannot generally solve the `test` environment on the bases of an already-solved `prod` environment by adding more packages.

Reply:
Why can't we first solve prod and then test with exact pinnings for the test env? Is it because this lowers the chance of the environment being solvable?

Reply:
Yes, whereas the other way round it's always solvable. I'll try to come up with a simple example

Reply:
Example:

- prod: `python`
- test: `python turbodbc`

If you solve prod first, you get `python=3.12`, which `turbodbc` doesn't support (as of today).

Reply:
This is indeed a different use case, which is not covered in the proposal. You want some kind of subtraction for an environment. Do you have an idea how to specify that?

Reply:
I would still say that this is the same use case but we haven't thought about the implications of when to solve what.

This will imply that a "prod" environment can only be extended by max one other environment, right?

Reply:
This should have been @pavelzw's "prod/test" use case I believe, because the "addition" way generally doesn't work. It fails for very simple use cases already, see the `python turbodbc` example

Reply:
I would say it does require a different notation as it changes the solve for the prod requirements which could be confusing. 

Maybe something like:
```toml
[feature.test]
dependencies = { pytest = "*" }
tasks = { test = "pytest" }

[dependencies]
python = "3.9"

[environments]
test = ["test"]
 # Keeps the solved environment just remove the dependencies (and tasks, channels, platforms, etc. it would add)
prod = { environment = "test", without-features = ["test"]}
```

Reply:
Hmm, I personally find the `without-features` weird and unintuitive to write 😅

I would prefer sticking with the original syntax as it's pretty easy to write and understand.
IMO if we document and communicate properly what's happening under the hood when you specify `test = {extends = "prod"}` and make clear that the solve of the prod environment will also be affected by this, this syntax should be fine.

If the user tries to extend an environment multiple times, we should raise an error.

```toml
[environments]
prod = []
test = {extends = "prod", features = ["test"]}
test2 = {extends = "prod", features = ["test", "py312"]}
```

```shell
$ pixi install
Error: cannot solve because the environment `prod` is extended multiple times.
<show occurrences in pixi.toml>
See https://pixi.sh/multiple-envs/#extending-envs why this isn't possible.
```

Reply:
> More generally, we are trying to solve two independent environments $A$ and $B$ in a way that they are compatible.

Just saw @jonashaag's edit.
With this intuition, we could solve this use-case with something like this instead

```toml
[environments]
prod = []
test = ["test"]
[environments.keep-compatible]
- ["prod", "test"]
```

(I'm not an expert in toml syntax, this is probably incorrect toml syntax but I hope you get my idea 😅)

I think this is also an interesting approach; don't have strong opinions which of these two approaches I find better.

Reply:
What do you think about this syntax:

```
[environments]
test = { name = "test", solve-group = "prod-deps", with-features = ["test"] }
prod = { environment = "test", solve-group = "prod-deps" }
```

This would signal the user very clearly, what envirnoments get combined into a big union for solving and how this feature works "under-the-hood".

Reply:
I really like the solve-group notation 🤩

Reply:
Yeah the solve group sounds good! Would that replace the `environment` field? I think the `with-` is not needed if there is no `without-`.

Reply:
> Would that replace the environment field?

Exactly. Can you update the docs accordingly?

Reply:
Changed it